### ISSUE TITLE
Roslyn doesn't handle inferred nullability in the presence of `goto`.

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#39220]
**Version Used**: 
```
Branch master (23 Sep 2019)
Latest commit 89b4f60 by Charles Stoner:
Allow conversion of collection initializer Add extension method this arg (#38732)
```

**Steps to Reproduce**:

Compile and run the following code:
```cs
#nullable enable
class C<T>
{
  public C(T t) => Property = t;
  public T Property { get; }
}

class Program
{
  static void Main()
  {
    Test("");
  }
  
  static void Test(string? s)
  {
    if (s == null) return;
    
    hell:
    var c = GetC(s);
    switch (c)
    {
      case { Property: var prop }:
        prop.ToString();
        break;
    }
    
    s = null;
    goto hell;
  }
  
  public static C<T> GetC<T>(T t) => new C<T>(t);
}
```

**Expected Behavior**:
`warning CS8602: Dereference of a possibly null reference.` reported for `prop.ToString()`

**Actual Behavior**:
No warnings are reported for the program above. The program crashes at runtime with a `NullReferenceException`

**Notes**
If you specify the type of `c` explicitly i.e. `C<string?> c = GetC(s);` you'll get the correct warning.

**Comment by gafter** [comment]
This has nothing to do with pattern-matching.  The following program also incorrectly reports no warnings:

``` c#
#nullable enable
class C<T>
{
    public C(T t) => Property = t;
    public T Property { get; }
}

class Program
{
    static void Main()
    {
        Test("");
    }

    static void Test(string? s)
    {
        if (s == null) return;

        hell:
        var c = GetC(s);
        var prop = c.Property;
        prop.ToString();
        s = null;
        goto hell;
    }

    public static C<T> GetC<T>(T t) => new C<T>(t);
}
```
